### ISSUE TITLE
Substantive changes to a competency - Ghosting

### ISSUE THREAD

**Issue by stuartasutton** [CredentialEngine/Schema-Development#563]
On occasion, there are significant changes to a published competency node that require a new URI. For example, let’s assume that significant, substantive edits have been done to competency node “B” in the graph below and that a new node with its own URI needs to be coined to replace the old node in its original graph position—i.e., the new node shares the same parent and child nodes with the old node.

ASN has discussed handling such changes through a process we’ve been calling “ghosting”—i.e., changing references in and to the original node from present to past tense thus maintaining familial memory of the ghost in the graph while adding its replacement node.

![2018-12-26 ghosts-a](https://user-images.githubusercontent.com/2939046/50562412-9425ab80-0cc8-11e9-8a5e-42cf3f9c4119.png)
The basic algorithm for such a change is embodied in the image below and includes the following steps:
 
1. clone node “B” to “B-1” and make the substantive changes to “B-1";
1. add a replaces object property to the new B-1 node that references the original node “B”;
1. add a replacedBy object property to the original “B” node to connect the old node to the new node;
1. edit the new "B-1" node to add a replaces object property referencing "B";
1. edit the parent node “A” to realign it’s hasChild object property to the URI of the new “B-1” node; 
1. edit the parent node “A” to add a hadChild object property to reference the old "B";
1. edit the child node “C” to align its isChildOf object property to the new “B-1” node; and
1. add a new wasChildOf object property node “C” to align it to its old parent, node B.

![2018-12-26 ghosts-b](https://user-images.githubusercontent.com/2939046/50562457-e797f980-0cc8-11e9-9cfa-97d79636fbfc.png)

**Note 1:** The ghosting in this case did not result in any significant substantive changes to either the parent node “A” or the child node “C”. Had the change to node “B” also involved significant substantive changes to either node “A” or “C” triggering new URIs for each of them, then ghosting would need to occur to those nodes as well. 

**Note 2:** Had the change been to the context of  node “B” by moving it to a different location in the graph, significant ghosting of the child nodes would likely be required.

**Comment by stuartasutton** [comment]
Nate, there's a significant difference between documenting change in a schema (e.g., CTDL-ASN, Dublin Core etc.) and handling changes in competency data (particularly, linked data).

**Comment by siuc-nate** [comment]
What I mean is, I think it would be far cleaner to just have the competency data link to some sort of history tracking mechanism (which itself might make use of such "past-tense" properties in its triples) rather than try to stuff the entire history of a competency into the data for that competency. Wouldn't that history also need to be reflected in every competency that was part of a change too? This would essentially mean that the same change has to be described three times (or its description split up into 3 places, which might actually be harder to track down - especially if one of those competencies is then moved/ghosted) if a competency is made the child of another competency, for example.

I think that can get very messy and may not scale well if there are a lot of changes or changes to a lot of competencies. It also doesn't have any information to indicate the timing or order of changes. For example, how would you indicate that Competency A was a child of Competency B, but before that was a child of Competency C, but before that was a parent of Competency B? Anyone concerned with the history of a competency is probably going to want to know when (or at least in what order) such changes were made.

I would leave it up to some external object to track this kind of archival information (whether such an object ends up looking like what we use for history tracking in CTDL or not, as long as it's outside of the data for the competency itself, it should be fine). This would also make it easier to indicate changes in groups so that the consumer of the data doesn't have to go fishing in every competency (past and present) in the entire framework in order to figure out its history.

It might be cleaner to just store previous versions of a framework and its competencies all together, as complete standalone objects (perhaps in parallel with the history tracking data), as this avoids needing to follow complex webs of changes in an attempt to rebuild a framework as it existed in some past state. Though this might have scaling problems too as it implies needing to duplicate the entire framework if one word is changed in one of its competencies.  There is probably a suitable middle ground somewhere that we should discuss.

One of the major issues a previous project we worked with/on had was the need to download not only the main object, but every bit of history attached to that object, in order to get a sense of what that object's final/complete data should be. This would be the same problem, but in reverse - I don't think we want to go there.

Perhaps the history could be referenced at the framework level? Then changes could still be grouped together by framework without needing to have every "edition" of a competency reference the history for that competency.  In the example below I try modeling this based more on the changes we use here in github rather than the history structure used by the schema management system:
```
//Framework - retrieved via https://framework/abc
"@graph": [
  {
    "@id": "https://framework/abc",
    "@type": "ceasn:CompetencyFramework",
    "ceasn:history": "https://history/abc"
  },
  {
    "@id": "https://competency/123",
    "@type": "ceasn:Competency",
    "ceasn:competencyText": "The current text for this competency"
  }
]

//History - retrieved via https://history/abc
"@graph": [
  {
    "@id": "https://history/abc",
    "@type": "history:ChangeLog",
    "history:changesFor": "https://framework/abc",
    "history:changeList": [
      "https://history/abc/19450824",
      "https://history/abc/023840923"
    ]
  }
  {
    "@id": "https://history/abc/19450824",
    "@type": "history:changeItem",
    "history:changeDate": "2019-01-02",
    "history:changeType": "history:removal"
    "history:changeData": {
      "rdf:subject": "https://competency/123",
      "rdf:predicate": "ceasn:hasChild",
      "rdf:object": "https://competency/456"
    }
  },
  {
    "@id": "https://history/abc/023840923",
    "@type": "history:changeItem",
    "history:changeDate": "2019-01-04",
    "history:changeType": "history:addition",
    "history:changeData": {
      "rdf:subject": "https://competency/123",
      "rdf:predicate": "ceasn:hasChild",
      "rdf:object": "https://competency/789"    
    }
  }
]
```
Just throwing that out there for discussion - I'm not currently proposing any of those properties or classes.  Note that if you wanted to have a property on each competency to reference its history, this would also work well with the above example.  Something like:
```
  {
    "@id": "https://competency/123",
    "@type": "ceasn:Competency",
    "ceasn:competencyText": "The current text for this competency",
    "ceasn:history": "https://history/for/https%3A%2F%2Fcompetency%2F123"
  }
```
Or some other similar structure that would **transparently** query for all history nodes with a subject or object that matches the provided `@id`.

@stuartasutton @Lomilar

**Comment by Lomilar** [comment]
I think we can all agree that there is no standard for how this type of thing should be handled, and they are all poor in one way or another. Here is another way, embedded with its own assumptions.

In CaSS, without using custom URLs (as we do in the CE context), we have "versioned URIs" and "unversioned (latest) URIs"

as has caused no end of headache for Stuart already:

`http://competency/<id>/<version>` - This URI refers to a specific competency that was saved at a specific time
`http://competency/<id>` - This URI refers to whatever the latest version of that competency is.

As anyone who comes from an RDF-first background knows, these two URIs are pointing at *different things* even though they may resolve (via an HTTP GET) to the same object. This breaks RDF processors, because the URI has to go through a system to resolve the alias.

They both may point at:
`{@id:"http://competency/<id>/<version>"}`

They may point at:
`{@id:"http://competency/<id>/<version>"}` and
`{@id:"http://competency/<id>/<newerVersion>"}`

Nonetheless, this gave us a flexibility that we didn't have previously, to either refer to the notion of a thing, updating over time, or to refer to a thing fixed in time. This is very important for people who want to refer to something to control. If I want to state you hold a competency, I want to state you hold this competency as defined at this particular time. But if I want to teach a competency today, then I want to teach the lastest definition of that competency.

We had to build rules about how to construct URIs and what they do (seen [here](http://devs.cassproject.org/index.html?doc=11HJPjgfxqKbCuBLaMuFLgtisCR-vyr8P89nsGyV-sD8))

and this does break quite a few RDF tools that expect references to a thing to be equal to the @id of the reference. Instead, we expect systems to compare to the object at the GETted reference's @id.

This assumption lets us also alias URIs, permitting the weird MD5'd references for external data, redirects,  etc.

But make no mistake, this is an intentional assumption that enables flexibility and reduces work and creates expectations. It could even be considered kinda sloppy or kinda elegant, depending on who you are.

You could use git to manage all the objects and have a way to say "This object with this commit", and then the consumer could just use all the git code management tools to manage their documents.

All viable, all with their own upsides and downsides.

**Comment by Lomilar** [comment]
Another disgusting and utterly practical and useable example of how to manage URIs can be found at:

https://docs.ipfs.io/guides/concepts/cid/

**Comment by Lomilar** [comment]
Another:

https://restfulapi.net/versioning/

Why you shouldn't do things the way we said:

https://www.troyhunt.com/your-api-versioning-is-wrong-which-is/

**Comment by jkitchensSIUC** [comment]
@stuartasutton @siuc-nate @Lomilar is this a Registry, CaSS, or CTDL Issue?

**Comment by Lomilar** [comment]
This can be resolved in quite a few places. cassproject/cass#499, for instance, is an application level resolution to the problem. Stuart would like something down a bit further at the schema or format level, so that a version of a competency can reference its previous version or reference what has changed.

In a nutshell, the web wasn't built to support multiple versions of data at the same URL, so there's a lot of possibilities around how we all decide to solve this.

**Comment by stuartasutton** [comment]
This is a CTDL issue (and actually a metadata issue in general). We veered way off base from how the issue was framed by me. The issue assumed that a full node in a graph was superseded--i.e., replaced with a node with a new URI. And, as might be expected: (1) URI versioning is a nonstarter for reasons discussed preciously, but also because we are not talking about versioning something rather superseding it.; and (2) no a (great) change management tracking based on changes to triples doesn't work because the substance of the superseded node doesn't change, it's superseded by something with a new URI and remains to be resolved as an entity.

**Comment by siuc-nate** [comment]
Per our 10-13-2020 meeting:

Closing this - it can be reopened later if needed.

**Comment by stuartasutton** [comment]
@siuc-nate @jeannekitchens @Lomilar, I am reopening this issue it may well be upon us.

**Comment by siuc-nate** [comment]
Archiving this, as it hasn't seen demand in a few years.
